### ISSUE TITLE
systemd_journald: Transform keys from journald JSON to Cloud Logging JSON

### ISSUE THREAD

**Issue by akkornel** [GoogleCloudPlatform/ops-agent#256]
Hello!

I am looking at using the Ops Agent for my GCP project, and I'm especially interested in using the `systemd_journald` receiver.  I am interested in this because journald is already collecting logs for me, and it would be great to take advantage of journald's structured-log format.

Even though the receiver is doing the work of pulling in logs from journald, the log entries coming from journald aren't in a form that Cloud Logging can process.  So, my request is that the Ops Agent's systemd_journald receiver be enhanced to transform the JSON entries from systemd's schema to Cloud Logging's schema.

One key from the JSON needs a simple key change:

* The key `MESSAGE` needs to change to `message`.

One key needs both a key change and a value change:

* `SYSLOG_PRIORITY` is a numeric priority encoded as a string.  The key name needs to change to `severity`, and the value needs to be mapped to an acceptable-to-Cloud-Logging string using the following mapping:
`7` maps to `DEBUG`
`6` maps to `INFO`
`5` maps to `NOTICE`
`4` maps to `WARNING`
`3` maps to `ERROR`
`2` maps to `CRITICAL`
`1` maps to `ALERT`
`0` maps to `EMERGENCY`

There are three keys which could be present and, if so, need to trigger the creation of the `logging.googleapis.com/sourceLocation` object:

* `CODE_FILE`: If present, it should be added to the `logging.googleapis.com/sourceLocation` object, under key `file`.
* `CODE_LINE`: If present, it should be added to the `logging.googleapis.com/sourceLocation` object, under key `line`.
* `CODE_FUNC`: If present, it should be added to the `logging.googleapis.com/sourceLocation` object, under key `function`.

Finally, the timestamp object has to be created by applying some math to the `__REALTIME_TIMESTAMP` key:

For the `timestampSeconds` key: Take the value from `⌊__REALTIME_TIMESTAMP ÷ 1000000⌋`.
For the `timestampNanos` key: Take the value from `__REALTIME_TIMESTAMP mod 1000000 × 1000`.

**Comment by quentinmit** [comment]
Severity mapping and code mapping are both good feature requests.

Are you actually having problems with the message and timestamp? Those should work fine without any additional translation.

**Comment by quentinmit** [comment]
I'm going to reopen this since it looks like we haven't fixed `MESSAGE` yet.

**Comment by alemnew** [comment]
Is there any ETA for this to be fixed and released?

**Comment by akkornel** [comment]
I don’t think this should be marked stale, as I do not think https://github.com/GoogleCloudPlatform/ops-agent/issues/256#issuecomment-1093315645 has been addressed.

**Comment by akkornel** [comment]
I don’t think this should be marked stale, as I do not think https://github.com/GoogleCloudPlatform/ops-agent/issues/256#issuecomment-1093315645 has been addressed.

**Comment by akkornel** [comment]
Hello!  I should have done this before, but I finally have submitted Google Cloud Support case #60294428, asking for someone to take a look at this.

**Comment by franciscovalentecastro** [comment]
Hello @akkornel ! Thank you for creating a Google Cloud Support case.

### Timestamps

I've explored the Systemd [^1] (footer link) receiver from Fluenb-bit (which is used as the current implementation the `systemd_journald`) and **found that the requested `timeStamp` transformations are already being done correctly** : 
```
For the timestampSeconds key: Take the value from ⌊__REALTIME_TIMESTAMP ÷ 1000000⌋.
For the timestampNanos key: Take the value from __REALTIME_TIMESTAMP mod 1000000 × 1000.
```
In [^2] (footer link) you can see the same calculations done in C code in the fluent-bit repository.

### `MESSAGE` to `message`

Since https://github.com/GoogleCloudPlatform/ops-agent/pull/502 also addressed several features in this issue, **as far as i understand the only missing feature is to transform `MESSAGE` to `message`** in the output LogEntry in Google Cloud Logging. 

We are currently exploring the task on how to incorporate this into the existing `systemd_journald` receiver. We have a workaround that can address this in the meantime.

### Workaround
This configuration can be used to transform `MESSAGE` -> `message` (comment https://github.com/GoogleCloudPlatform/ops-agent/issues/256#issuecomment-1093315645) in the resulting logs  :
```
logging:
  receivers:
    systemd_logs:
      type: systemd_journald
  processors:
    modify:
      type: modify_fields
      fields:
        jsonPayload.message:
          move_from: jsonPayload.MESSAGE
  service:
    pipelines:
      systemd_pipeline:
        receivers: [systemd_logs]
        processors: [modify]
```

[^1]: https://docs.fluentbit.io/manual/pipeline/inputs/systemd
[^2]: https://github.com/fluent/fluent-bit/blob/master/plugins/in_systemd/systemd.c#L336-L348
